### PR TITLE
Use StateLocker in MinMaxSumCount

### LINKED ISSUE #437: Some aggregators may have inconsistent states when updating multiple values
`MinMaxSumCount` and `Histogram` aggregators keeps multiple fields that are independently atomically updated by `Updated()` and `Checkponit()`, which could result in inconsistent states when these two methods are called concurrently. This is a result of the lock-free implemented algorithm on these aggregators.

From a study based on the Prometheus implementation that @jmacd linked in the #385, we could achieve a lock-free `Update()` algorithm that is consistent by making `Checkpoint()` a blocking method, since the last is called by the SDK itself this should not be a problem for the end user.

[Link to the mentioned algorithm](https://github.com/prometheus/client_golang/blob/e7776d2c54305c1b62fdb113b5a7e9b944c5c27e/prometheus/histogram.go#L227). 
[Link to the discussion](https://github.com/open-telemetry/opentelemetry-go/pull/385/files#r358592144).

### LINKED ISSUE #535: Use StateLocker in MinMaxSumCount
#438 introduces a new `StateLocker` type. 

The MinMaxSumCount aggregator should take advantage of it.

### PR DESCRIPTION
**Pull request by evantorrie**
Closes #535 
Closes #437 

   * Reimplement MinMaxSumCount using StateLocker to ensure a consistent set of values for the aggregator's outputs.

   * Added a stress test that fails under the earlier MinMaxSumCount implementation, but succeeds with the new StateLocker-based implementation.

### CODE REVIEW

File: sdk/metric/minmaxsumcount_stress_test.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
+// Copyright 2019, OpenTelemetry Authors
```
Comment (MrAlias) [inline sdk/metric/minmaxsumcount_stress_test.go]:
```suggestion
// Copyright 2020, OpenTelemetry Authors
```

File: sdk/metric/minmaxsumcount_stress_test.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	"go.opentelemetry.io/otel/api/core"
+	"go.opentelemetry.io/otel/sdk/export/metric"
+	"go.opentelemetry.io/otel/sdk/metric/aggregator/minmaxsumcount"
+)
+
+func TestStressInt64MinMaxSumCount(t *testing.T) {
+	desc := metric.NewDescriptor("some_metric", metric.MeasureKind, nil, "", "", core.Int64NumberKind)
+	mmsc := minmaxsumcount.New(desc)
+
+	go func() {
```
Comment (MrAlias) [inline sdk/metric/minmaxsumcount_stress_test.go]: This is going to leak. Probably want to include an end signal. Given the context you need to pass to `Update` you could pass an external context in with a cancel signal.

Not the end of the world given the short life of the test harness, but something to consider.

File: sdk/metric/minmaxsumcount_stress_test.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	desc := metric.NewDescriptor("some_metric", metric.MeasureKind, nil, "", "", core.Int64NumberKind)
+	mmsc := minmaxsumcount.New(desc)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
+		rnd := rand.New(rand.NewSource(time.Now().Unix()))
+		v := rnd.Int63() % 103
+		for {
+			_ = mmsc.Update(ctx, core.NewInt64Number(v), desc)
```
Comment (MrAlias) [inline sdk/metric/minmaxsumcount_stress_test.go]:
```suggestion
			select {
			case <-ctx.Done():
				return
			default:
				_ = mmsc.Update(ctx, core.NewInt64Number(v), desc)
			}
```

File: sdk/metric/minmaxsumcount_stress_test.go (hunk 2 above)
Comment (MrAlias) [reply sdk/metric/minmaxsumcount_stress_test.go]: @paivagustavo looping for a definite amount of time could work, but I like returning with the signal included in the context. The context is already going to be there and it doesn't waste any more computation time than needed.

Though, again, doesn't need to be the most resource conservative given where this lives.

File: sdk/metric/minmaxsumcount_stress_test.go (hunk 3 above)
Comment (evantorrie) [reply sdk/metric/minmaxsumcount_stress_test.go]: Sorry, don't know what I was thinking on my first attempt...

File: sdk/metric/aggregator/minmaxsumcount/mmsc.go [hunk 4]
```diff
@@ -48,104 +48,116 @@ var _ aggregator.MinMaxSumCount = &Aggregator{}
...
 
 // Min returns the minimum value in the checkpoint.
-// The error value aggregator.ErrEmptyDataSet will be returned if
-// (due to a race condition) the checkpoint was set prior to
-// current.min being computed in Update().
-//
-// Note: If a measure's recorded values for a given checkpoint are
-// all equal to NumberKind.Maximum(), Min() will return ErrEmptyDataSet
+// The error value aggregator.ErrEmptyDataSet will be returned
+// if there were no measurements recorded during the checkpoint.
```
[the remark this answers is not in the document]
Comment (evantorrie) [reply sdk/metric/aggregator/minmaxsumcount/mmsc.go]: Doc already has `Min returns the ...` line above these two.

File: sdk/metric/aggregator/minmaxsumcount/mmsc.go [hunk 5]
```diff
@@ -48,104 +48,116 @@ var _ aggregator.MinMaxSumCount = &Aggregator{}
 // New returns a new measure aggregator for computing min, max, sum, and
 // count.  It does not compute quantile information other than Max.
 //
-// Note that this aggregator maintains each value using independent
-// atomic operations, which introduces the possibility that
-// checkpoints are inconsistent.  For greater consistency and lower
-// performance, consider using Array or DDSketch aggregators.
+// This aggregator uses the StateLocker pattern to guarantee
+// the count, sum, min and max are consistent within a checkpoint
```
[the remark this answers is not in the document]
Comment (evantorrie) [reply sdk/metric/aggregator/minmaxsumcount/mmsc.go]: Paragraph does start with `New returns a new measure aggregator...`. In this PR, I just changed two lines that were already part of the godoc

File: sdk/metric/minmaxsumcount_stress_test.go (hunk 1 above)
Comment (evantorrie) [reply sdk/metric/minmaxsumcount_stress_test.go]: I'll do that (and the goroutine leak) in a separate PR

File: sdk/metric/minmaxsumcount_stress_test.go (hunk 2 above)
Comment (evantorrie) [reply sdk/metric/minmaxsumcount_stress_test.go]: I used the `context.WithCancel(..)` idiom and `defer cancelFunc()`.

File: sdk/metric/minmaxsumcount_stress_test.go (hunk 1 above)
Comment (evantorrie) [reply sdk/metric/minmaxsumcount_stress_test.go]: Created PR #547 relating to copyright date and goroutine leak in histogram stress test.

File: sdk/metric/minmaxsumcount_stress_test.go (hunk 3 above)
Comment (MrAlias) [reply sdk/metric/minmaxsumcount_stress_test.go]: no worries 😄

Comment (MrAlias) [reply sdk/metric/minmaxsumcount_stress_test.go]: No worries 😄

**Pull request merged**